File: Packages/StackEngineCore/Sources/StackEngineCore/BoxFilter.swift
/// Separable window mean (edge-clamped, normalized by the true in-image sample count at borders).
/// The smoothing primitive behind the guided filter.
enum BoxFilter {
    static func mean(_ src: [Float], width w: Int, height h: Int, radius r: Int) -> [Float] {
        var tmp = [Float](repeating: 0, count: w * h)   // horizontal pass
        for y in 0..<h {
            for x in 0..<w {
                var s: Float = 0; var n = 0
                for dx in -r...r { let xx = x + dx; if xx >= 0, xx < w { s += src[y * w + xx]; n += 1 } }
                tmp[y * w + x] = s / Float(n)
            }
        }
        var out = [Float](repeating: 0, count: w * h)   // vertical pass
        for y in 0..<h {
            for x in 0..<w {
                var s: Float = 0; var n = 0
                for dy in -r...r { let yy = y + dy; if yy >= 0, yy < h { s += tmp[yy * w + x]; n += 1 } }
                out[y * w + x] = s / Float(n)
            }
        }
        return out
    }
}

File: android/stackengine/src/main/kotlin/com/jdpneto/stackengine/BoxFilter.kt
package com.jdpneto.stackengine

/**
 * Separable window mean (edge-clamped, normalized by the true in-image sample count at borders).
 * The smoothing primitive behind the guided filter.
 */
internal object BoxFilter {

    /**
     * Compute the box-filter mean of [src] with the given half-window [radius].
     *
     * Two-pass separable implementation (horizontal then vertical). Border pixels are handled
     * by edge-clamping: only in-image samples contribute and the normalization denominator
     * counts only those samples (so border values are the mean of real samples, not darkened).
     */
    fun mean(src: FloatArray, w: Int, h: Int, r: Int): FloatArray {
        val tmp = FloatArray(w * h)    // horizontal pass result
        for (y in 0 until h) {
            for (x in 0 until w) {
                var s = 0f
                var n = 0
                for (dx in -r..r) {
                    val xx = x + dx
                    if (xx in 0 until w) { s += src[y * w + xx]; n++ }
                }
                tmp[y * w + x] = s / n.toFloat()
            }
        }
        val out = FloatArray(w * h)    // vertical pass result
        for (y in 0 until h) {
            for (x in 0 until w) {
                var s = 0f
                var n = 0
                for (dy in -r..r) {
                    val yy = y + dy
                    if (yy in 0 until h) { s += tmp[yy * w + x]; n++ }
                }
                out[y * w + x] = s / n.toFloat()
            }
        }
        return out
    }
}
